blueprints/varnish-support.md

```diff
@@ -0,0 +1,116 @@
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+# Add Varnish Cache Support
+
+## Problem Description
+
+Currently Traffic Control uses Traffic Server as the underlying cache server. We can expand that by introducing Varnish cache as an option for the cache server used with its great performance, robustness and modularity.
+
+## Proposed Change
+
+From a high level point of view, ATS operates based on configuration files that describe in details how it should work and interact with other servers in the cache hierarchy. These configuration files are managed and generated using `t3c` components that utilize Traffic Ops APIs to get profiles and parameters data required for the configuration files. The proposed change is to use the same data fetched from Traffic Ops APIs to generate configuration files for Varnish cache with almost the same functionality.
+
+Note: the changes should not affect existing components but rather build on them.
+
+### Traffic Portal Impact
+
+n/a
+
+### Traffic Ops Impact
+
+n/a
+
+#### REST API Impact
+
+n/a
+
+#### Client Impact
+
+n/a
+
+#### Data Model / Database Impact
+
+- The profile type `ATS_PROFILE` will be renamed to `CACHE_PROFILE` to indicate that the profile is used for any cache server not just ATS while parameters and other fields won't be affected.
+- `DeliveryService` structs contain fields related to ATS like `remapText`. It will be parsed and translated to Varnish configuration.
+
+### ORT Impact
+
+- `varnishcfg` package will be developed to handle generating configuration files for Varnish, Hitch and `varnishncsa`. For detailed description of mapping configuration files from ATS to Varnish refer to [Varnish Support](https://github.com/apache/trafficcontrol/wiki/Varnish-Support) wiki.
+- New options will be added to `t3c-generate` and `t3c-apply` including `--cache` to indicate which cache server the configuration files will be generated or applied to (e.g. `--cache=varnish` or `--cache=ats`). Flags will be rewritten to indicate which cache server they can be used with. `t3c` subcommands will decide based on `cache` option whether to use `go-atscfg` or `varnishcfg` for configuration files generation and also how to apply them for each case.
```

Comment:
It could be done this way or a server profile parameter could be used to indicate what type of server it is.  Server profiles already reference ATS specific rpm packages (although t3c only supporting redhat type distros is probably bad).

Reply:
I strongly oppose using the loosey-goosey free-form text of a Parameter to define things like this. If it's desirable that `t3c` automatically detect the cache server software, and that can't be done by inspecting the system locally (which would be ideal but difficult), then I'd vote on bringing back the Profile Type that's specific to Varnish cache servers.

Reply:
those options seem reasonable.